docker/README.md

```diff
@@ -0,0 +1,109 @@
+images for running the various contract families in the PDO contracts
+repository. The images may be used for a variety of uses including
+demonstration, testing and interacting with PDO contracts. The second
+purpose is to provide a means for clean, reproducible testing of the
+contracts and the Jupyter notebooks. Fully automated testing
+(e.g. automated execution of all contract family tests) is provided
+through the `test` target in the `Makefile` (e.g. run `make test` to
+perform the automated tests). Interactive testing of the Jupyter
+notebooks is provided through the `test_jupyter` target. That target
+will create and run a container with a Jupyter server running inside.
+Note that this jupyter services is also a good place to get some some hands-on experience with the mechanics of PDO contrarts: after running `make test_jupyter`, peruse the [documentation](../exchange-contract/docs/notebooks/README.md) and point your browser to `http://localhost:8888` -- note jupyter listens only to localhost, if you are running this on a headless remote server accessed via ssh, just forward the port from your local machine with `ssh $YOUR_TEST_MACHINE -L8888:localhost:8888` ... -- to experiment.
+
+## Basic Layout ##
+
+There are three subdirectories that are employed in building, configuring and running a container.
+
+* `xfer` -- this directory is used to pass configuration information
+  and keys between the container and the host; for example, to push a
+  previously built configuration into the container, put the files in
+  the appropriate subdirectory in xfer.
+* `tools` -- this directory contains a number of scripts that will be
+  installed in the container to simplify building, configuring and
+  running the services in the container.
+* `repository` -- this directory is created during the build process
+  and contains the PDO contracts source code that will be copied into
+  the container; the build variables `CONTRACTS_REPO` and
+  `CONTRACTS_BRANCH` control what is put into the directory.
+
+## Makefile Targets ##
+
+The `Makefile` contains several targets that should simplify building
+images and running containers.
+
+Four configuration variables should be set as necessary:
+
+* `CONTRACTS_REPO` -- the URL or file path of the PDO source git
+  repository; this defaults to the repository stored on the local file
+  system in which the docker directory is located.
+* `CONTRACTS_BRANCH` -- the branch to use in from the source repository;
+  this defaults to the branch where the source is currently stored.
+* `CONTRACTS_USER_UID` -- the UID for the user that is created in the
+  container to run the services; this defaults to the current users
+  UID. Note that the `xfer` directory must be writable by the account
+  associated with the UID.
+* `CONTRACTS_GROUP_UID` -- the GID for the group assigned to the user
+  created in the container; the default is the GID for the current
+  user.
+
+### Testing PDO Contracts  ###
```

Comment:
There are really two parts to this sub-section: Running automated tests and setting up a clean Jupyter notebook environment for interactive testing. As above, breaking these paragraphs down by purpose would help guide the reader towards their intended action.